modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala

```diff
@@ -1523,6 +1523,101 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
+            |def foo(): Int = 23
+            |
+            |def test: Int = foo
+            |//              ^^^ error: missing parentheses
+            |""".stripMargin
+      )
+
+      withBsp(inputs, Seq(".")) { (root, localClient, remoteServer) =>
+        async {
+          val buildTargetsResp = await(remoteServer.workspaceBuildTargets().asScala)
+          val target = {
+            val targets = buildTargetsResp.getTargets.asScala.map(_.getId).toSeq
+            expect(targets.length == 2)
+            extractMainTargets(targets)
+          }
+
+          val targetUri = TestUtil.normalizeUri(target.getUri)
+          checkTargetUri(root, targetUri)
+
+          val targets = List(target).asJava
+
+          val compileResp = await {
+            remoteServer
+              .buildTargetCompile(new b.CompileParams(targets))
+              .asScala
+          }
+          expect(compileResp.getStatusCode == b.StatusCode.ERROR)
+
+          val diagnosticsParams = {
+            val diagnostics = localClient.diagnostics()
+            val params      = diagnostics(2)
+            expect(params.getBuildTarget.getUri == targetUri)
+            expect(
+              TestUtil.normalizeUri(params.getTextDocument.getUri) ==
+                TestUtil.normalizeUri((root / "test.sc").toNIO.toUri.toASCIIString)
+            )
+            params
+          }
+
+          val diagnostics = diagnosticsParams.getDiagnostics.asScala.toSeq
+          expect(diagnostics.length == 1)
+
+          val theDiagnostic = diagnostics.head
+
+          checkDiagnostic(
+            diagnostic = theDiagnostic,
+            expectedMessage =
+              "method foo in class test$_ must be called with () argument",
+            expectedSeverity = b.DiagnosticSeverity.ERROR,
+            expectedStartLine = 3,
+            expectedStartCharacter = 16,
+            expectedEndLine = 3,
+            expectedEndCharacter = 19
+          )
+
+          // Shouldn't this be set to "scala"? expect(theDiagnostic.getDataKind == "scala")
+
+          assertNoDiff(
+            theDiagnostic.getData.toString,
+            """{
+              |"actions":[
+              | {
+              |   "title": "Insert_()",
+              |   "edit": {
+              |     "changes": [
+              |       {
+              |         "range": {
+              |           "start": {
+              |             "line": 11,
+              |             "character": 19
+              |           },
+              |           "end": {
+              |             "line": 11,
+              |             "character": 19
+              |           }
+              |         },
+              |         "newText": "()"
+              |       }
+              |     ]
+              |   }
+              |}
+              |]
+              |}""".stripMargin
+              .replaceAll("\\s", "")
+              .replaceAll("_", " ")
```

Comment:
Can you try to deserialize this JSON object into `bsp4j.ScalaDiagnostic` as [here](https://github.com/VirtusLab/scala-cli/blob/5dd3559719e9c54498061b40b0a8055764fcf709/modules/build/src/test/scala/scala/build/tests/BuildTests.scala#L378)?. You can use Gson for the deserialization.